simplify api macro

scgilardi opened the pull request:
I'm still exploring what's changed in entitydb for v2. I had some ideas about the `export-api` macro which I've included here. Also I have some questions below about metadata on a named item. I'll be working with v2 on my current project to try to generate some useful feedback for you. Looks good so far.

- inline defns
  - no need for wrap-* fns
- corrects a couple of api fn names
- adds tests

Questions:
- the last test shows meta on the item is nil. is that correct?
- what's the correct way to retrieve meta?
- could we return meta as ClojureScript meta on the returned item?

retro commented:
@scgilardi I'm ok with exposing meta data on object's meta too. IIRC the problem with it is that changing meta is not triggering reagent's render cycle, which is why we need another way to expose it. 

Also, we have an unsolved issue with the exposing API, which I'm not sure how to approach:
In some cases (this is the case when wrapping Relay GraphQL API), you'll want to do some cleanup after item removal. For instance, when you remove the item that is a relation, you might want to remove the wrapper object too - https://github.com/retro/entitydb-v2/blob/master/test/entitydb/data.cljs#L53-L55 - in this case when the post with the slug `my-post-3` is removed the wrapper object will be left (the one with the `cursor` and node `attrs`). Expected behavior would be to remove that object too. Our idea is to write a custom function that would call `remove-by-id*` (https://github.com/retro/entitydb-v2/blob/master/src/entitydb/entitydb.cljs#L233), and then do the cleanup based on the report - which will contain all the affected paths, but I'm not sure how to expose that custom `remove` function through the `export-api` macro.

scgilardi commented:
@retro,
re: "IIRC the problem with it is that changing meta is not triggering reagent's render cycle, which is why we need another way to expose it.",

What occurs to me is that perhaps that's a good thing. I'm not sure what you envision (or have experience with) as the purpose for the metadata one can specify with as the `named-meta` argument to `insert-named-item`, but one reasonable role for it would be one similar to clojure metadata: "data you want to save with the item, but that shouldn't affect its value". Following from that, a reasonable rule could be: "if metadata changes, the value didn't change, so there was no re-render". If one wanted another kind of metadata that *should* cause a re-render if changed, a `value` for a `:meta` key (or `:some-namespace/meta`) could be used instead.

re: "For instance, when you remove the item that is a relation, you might want to remove the wrapper object too - https://github.com/retro/entitydb-v2/blob/master/test/entitydb/data.cljs#L53-L55 - in this case when the post with the slug my-post-3 is removed the wrapper object will be left (the one with the cursor and node attrs)",

With limited knowledge of how all this is used, I wonder why this "needs cleanup" relationship is different from a normal entity-db `1:1` relationship where if A refers to B, then when A is removed, B is also removed. or is that only done in a '1:many' relationship? Thinking in terms of an author owning articles and when the owner is removed the articles are also deleted. Maybe there could be different relationships like "1:1 related" (no implied deletions) and "1:1 owned" (delete the dep if the owner is deleted). If you have a more detailed example handy for the desired behavior and why existing cleanup is not sufficient, I'd appreciate seeing it so I can understand the problem better. I'll also read some more "on delete cascade" in SQL. So far I kind of like the idea of an expanded set of relationships that provide different lifetime (ownership) behaviors as a general solution in this area.

retro commented:
@scgilardi 
For the metadata usage (and why we _might_ want to re-render when it changes) - one of the usecases is tracking of item's status - for instance if we're loading it at the moment. We might want to show the spinner or some other indication to the user.

For the cleanup when the item is removed - If you remove item explicitly from the storage, it means that it won't be accessible from anywhere else in the data graph. Imagine that you have an email client with a master - detail view. If you delete an item by clicking the delete button in the detail panel, you would expect it to disappear from the list too. This makes writing consistent UIs easier. That part is already implemented. My question was related to the cases when you want to do some additional cleanup when this happens. In case of GraphQL (Relay) APIs, each entity in the list is wrapped with another object that holds it's cursor information (that can be used to paginate through the list). This object is only valid for that item, and we would like to not leave it in the system. This is why we've prepared the `remove-by-id*` function that returns new store state and the report of all places where we had to remove reference to the removed object. I'm wondering how could we have a custom remove-by-id function that is exposed by the macro and wrapped with the same behavior as the default functions. I would like something that could look like this (pseudo code, very bad API design):

```
(defn custom-remove-by-id-fn [...])
(edb/def-adapted-api edb-path :remove-by-id custom-remove-by-id-fn)
```

So I'm wondering how the API should look like, and how the implementation would work
